dataclients/kubernetes/crd.md

```diff
@@ -0,0 +1,898 @@
+# RouteGroup Proposal
+
+Date: 2019-09-04
+Status: open
+Owners: @aryszka @szuecs
+
+This proposal will be kept up to date, if you like to discuss it
+please use https://github.com/zalando/skipper/issues/660
+
+Disclaimer:
+Skipper's `kubernetes` dataclient will also support Ingress in the
+future.
+
+## Purpose
+
+The purpose of the document is to define a RouteGroup CRD, that allows
+configuring Skipper for Kubernetes such that it benefits from all the
+Skipper features in a straightforward way.
+
+## Problem statement
+
+At Zalando we have a controller, that creates multiple Ingresses for a
+given enhanced OpenAPI spec. It configures Skipper features with
+[predicates](https://opensource.zalando.com/skipper/reference/predicates)
+and
+[filters](https://opensource.zalando.com/skipper/reference/filters/)
+to allow different rate limits and permissions for different
+clients. This works great, but has weaknesses in case of traffic
+switching. Right now, traffic switching is orchestrated by
+`zalando.org/backend-weights` Ingress annotation.
+
+For example an Ingress object specifies traffic weights for service
+`my-app-1` with 80% of the traffic, rest will be routed to
+`my-app-2`:
+
+```yaml
+metadata:
+  name: my-app
+  labels:
+    application: my-app
+  annotations:
+    zalando.org/backend-weights: |
+      {"my-app-1": 80, "my-app-2": 20}
+spec:
+  rules:
+  - host: my-app.example.org
+    http:
+      paths:
+      - backend:
+          serviceName: my-app-1
+          servicePort: http
+        path: /
+      - backend:
+          serviceName: my-app-2
+          servicePort: http
+        path: /
+```
+
+The problem with having multiple Ingress having this kind of
+annotation is that it needs to be in-sync across all Ingress belonging
+to a group of routes. This has of course a race condition, which is a
```

Comment:
It's not clear why their is a race condition in this case. I actually don't know what the race condition is?

Reply:
race condition: multiple, not global synchronized, writes and a list read